### PR TITLE
Generate: speculative decoding

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Useful context:
In a recent PR (#27750), the candidate generation in assisted generation got abstracted, so we can host new candidate generation techniques (such as #27722).

_______________________________________________________

This PR:
1. ~Reworks assisted candidate generation to call `.generate()`, instead of having its own custom generation loop. For most models this is nothing more than a nice abstraction. However, for models with a custom `generate()` function, this means the assistant model will now make use of it! (🤔 does this mean that DistilWhisper gets better numbers with this refactor?)~ Edit: moved to #28030 
2. Adds speculative decoding ([paper](https://arxiv.org/pdf/2211.17192.pdf), see Algorithm 1). This implied a minor interface change in the candidate generation class, which should be okay since it hasn't been released :)

The following tests were run locally and are passing:
1. `RUN_SLOW=1 py.test tests/models/whisper/ -k speculative`
2. `py.test tests/ -k test_assisted` (which now triggers speculative decoding)

________________________________________________________

TODO:
- [ ] Benchmark speculative decoding

### CODE REVIEW

**Comment by gante** [comment]
@patrickvonplaten tagging you here for a 2nd set of eyes on the speculative decoding method (changes in `utils.py`), which I'm assuming you're familiar with. Feel free to delegate to someone else who is familiar with the method! 🤗

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4634]: These are not the best variable names, but it's hard to compare against the original algorithm if they don't match 🤔 As such, I've decided to keep the original names

**Comment by amyeroberts** [comment]
Thanks for adding this! Can we split this up into two separate PRs: one changing the assisted generation and the other adding speculative decoding?

**Comment by gante** [comment]
@amyeroberts pulled the assisted generation changes into this PR: https://github.com/huggingface/transformers/pull/28030

After it is merged, I will rebase this one -- this one will become exclusively about speculative decoding 🤗

**Comment by gante** [comment]
@amyeroberts I've rerun the slow tests, and I can confirm that they are passing. Ready for a review :)

File: src/transformers/generation/utils.py
Comment (amyeroberts) [inline src/transformers/generation/utils.py:4650]: We can take advantage of there being a comment here to explain that all tokens which have a higher probability from the assistant model are selected etc. otherwise (speaking for myself here) the user has to scroll back up to remember what p and q represent

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Can we add some tests, in particular one which checks case 1. and one which makes sure the correct logic branch is being selected e.g. checking candidate_logits is None when expected (might be a test on the candidate generator instead)?

File: src/transformers/generation/utils.py
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:4683]:
```suggestion
                if do_sample:
                    probs = new_logits.softmax(dim=-1)
                    selected_tokens = torch.multinomial(probs[0, :, :], num_samples=1).squeeze(1)[None, :]
                else:
                    selected_tokens = new_logits.argmax(dim=-1)
```

It's probably time to soon factor this out into something like:

```py
selected_tokens = Categorical(new_logits / temperature).sample()
```

everywhere

File: src/transformers/generation/utils.py
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:4680]: is this case still relevant? Not sure it's a good idea to have two "assisted decoding" do_sample=True cases in our generate. Should we maybe just deprecate this case?

File: src/transformers/generation/utils.py
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:4686]:
```suggestion
                candidate_new_tokens = candidate_input_ids[:, -candidate_length:]
                n_matches = ((~(candidate_new_tokens == selected_tokens[:, :-1])).cumsum(dim=-1) < 1).sum()
```

Just to be sure - does this have the correct indention - it looks like this is currently only used for Case 2 (but if that's intended, all good!)

File: src/transformers/generation/utils.py
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:4673]:
```suggestion
                # Ensure we don't generate beyond max_len or an EOS token (not in algorithm 1, but needed for correct
                # behavior)
                if last_assistant_token_is_eos and n_matches == candidate_length:
                    n_matches -= 1
                n_matches = min(n_matches, max_len - cur_len - 1)

                # Next token selection: if there is a rejection, adjust the distribution from the main model before
                # sampling.
                gamma = candidate_logits.shape[1]
                p_n_plus_1 = p[:, n_matches, :]
                if n_matches < gamma:
                    q_n_plus_1 = q[:, n_matches, :]
                    p_prime = torch.clamp((p_n_plus_1 - q_n_plus_1), min=0).softmax(dim=-1)
                else:
                    p_prime = p_n_plus_1
                t = torch.multinomial(p_prime, num_samples=1).squeeze(1)[None, :]

                # The selected tokens include the matches plus the next sampled token
                selected_tokens = torch.cat((candidate_input_ids[:, :n_matches], t), dim=-1)
```

(nit) wonder if we could factor this out into a function - it's quite hard to read and might be easier to just be put into a separate function

**Review by patrickvonplaten — COMMENTED** [review]
Super cool addition! 

Not really related to this PR, but I feel like we should start putting all the generation submethods (assisted decoding, greedy & sample (guess we can merge these two), beam search, ...) into their own files by now

My only important comment here is that I don't think it's great that we have 2 assisted generation cases now where `do_sample=True`. Can we deprecate the "non-official" one?

**Comment by gante** [comment]
@patrickvonplaten the two types of sampling are needed :D 

New candidate-based methods are popping up (e.g. https://github.com/huggingface/transformers/pull/27775), and they don't necessarily have logits. As such, speculative decoding, which needs the candidates' logits, can't be applied to those methods.

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten the two types of sampling are needed :D
> 
> New candidate-based methods are popping up (e.g. #27775), and they don't necessarily have logits. As such, speculative decoding, which needs the candidates' logits, can't be applied to those methods.

But shouldn't they just be the "own" method now? I.e. I don't think we should put https://github.com/huggingface/transformers/pull/27775 into the speculative decoding method no?

**Comment by gante** [comment]
@patrickvonplaten #27775 does not introduce changes to assisted generation 🤗  in #28030 I've abstracted the candidate generation part of assisted generation. We now load candidate generators the same way as we load the logits processors:

https://github.com/huggingface/transformers/blob/e6dcf8abd6f65bb4b6dfc1831b20d9ba49ce00e2/src/transformers/generation/utils.py#L899-L919

In assisted generation, we call the candidate generator to get candidate sequences (which may or may not contain associated logits, depending on the method)

https://github.com/huggingface/transformers/blob/e6dcf8abd6f65bb4b6dfc1831b20d9ba49ce00e2/src/transformers/generation/utils.py#L4588

The technique in #27775 can thus be added by adding a new candidate generator in ` _get_candidate_generator`. Other candidate generators may be added the same way, enabling users to experiment with the concept of candidates!

File: src/transformers/generation/utils.py
Comment (gante) [reply src/transformers/generation/utils.py:4683]: Yes! Then equivalent sampling/non-sampling methods (e.g. greedy decoding/samplinh) could be merged into a single function, facilitating maintenance. I'm going to leave it to a follow-up PR, though, to keep this PR exclusively about speculative decoding.

File: src/transformers/generation/utils.py
Comment (gante) [reply src/transformers/generation/utils.py:4686]: Yes, I confirm that it is the intended indentation 🤗

**Comment by gante** [comment]
@amyeroberts PR comments addressed 🤗 

@patrickvonplaten Unless you don't strongly oppose, I'd like to keep the two sampling paths, for the reasons I've written [here](https://github.com/huggingface/transformers/pull/27979#issuecomment-1860125398) -- I think it will be beneficial in the long run! :)

**Comment by gante** [comment]
@amyeroberts -- @patrickvonplaten and I had a chat about whether to keep the two sampling paths or not. For context, here's what we agreed on:
- It's okay to leave it as is, and perhaps abstract the different ways we accept candidates into a `candidate_checker` block. 
- Be conservative on adding new candidate generators, so we don't end up with unused methods
- [in a follow-up PR] squash other cases where the decoding method is the same except for the token selection, like `greedy_decoding` + `sample`
- [in a follow-up PR] mode each decoding method into its own file

**Review by patrickvonplaten — APPROVED** [review]

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating!

**Comment by jmamou** [comment]
@gante 
According to experiments reported in Leviathan's paper, speculative decoding (SD) has higher speedup with greedy decoding (temp=0). However, in the current implementation, SD works only with do_sample=True.

**Comment by gante** [comment]
@jmamou speculative decoding with `do_sample=False` (or `temp=0`) was already encoded in `assisted_generation` -- try calling `model.generate(input_ids, do_sample=False, assistant_model=assistant_model)` :)

**Comment by jmamou** [comment]
@gante 
Since acceptance criteria are different between speculative decoding and assisted generation, I think that it would be great to be able to run both speculative decoding and assisted generation with no sampling.

**Comment by jmamou** [comment]
@gante 
I implemented it. I can submit a PR.

**Comment by gante** [comment]
@jmamou 

> Since acceptance criteria are different between speculative decoding and assisted generation, I think that it would be great to be able to run both speculative decoding and assisted generation with no sampling.

Not sure if this is a good idea
1. if we see greedy decoding as applying `temperature=0`, the model probability will be `1` at the most likely token and `0` everywhere else. In turn, this implies that `p_i/q_i` is `>=1` at all positions, and thus all candidate tokens would be accepted 👉 speculative decoding would be the same as simply using the assistant model 
2. If we don't apply `temperature=0`, then it would be sampling -- in other words, it wouldn't be greedy decoding

> In previous implementation of assisted generation (4.33) with heuristical update of num_assistant_tokens (or max_assistant_tokens), the value of num_assistant_tokens was preserved between 2 consecutive generate() calls.

In current implementation (4.38), num_assistant_tokens is updated by the candidate_generator during the generation but assistant_model.generation_config.num_assistant_tokens is not updated at the end of the generation. Therefore, next call to generate will start with the initial value of assistant_model.generation_config.num_assistant_tokens (5).

Is it intentional? If that's a bug, I can open a PR to fix it.

This is a good point! A PR to revert to the previous behaviour (with a test) would be appreciated 🙏

**Pull request merged**
